# Bibliotecas

In [260]:
import os
import pandas as pd
import numpy as np
import scipy as sp
from scipy import signal
import timeit
from datetime import datetime
from sklearn.externals import joblib
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn import preprocessing
from sklearn import utils

from ml_time_series import generate_envelope

%matplotlib notebook

In [233]:
def HStacker(files_,t):
    
    #Lê dados do arquivo e remove janela
    Xc = np.load('DB_REPO/preproc/'+files_[3])
    Xc = Xc[:-49] #Exclui do STD para igualar o Hshape
    Xci = np.load('DB_REPO/preproc/'+files_[0])
    Xci = Xci[:-49]
    Xc = np.hstack((Xc.reshape(Xc.shape),Xci.reshape(Xci.shape)))
    
    #Aplica Hstack nos dados
    for j in range(1,len(files_)-1):
        Xci = np.load('DB_REPO/preproc/'+files_[j])
        Xc = np.hstack((Xc.reshape(Xc.shape),Xci.reshape(Xci.shape)))
          
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    np.save('./DB_REPO/stacked/'+t+'_hs', Xc, allow_pickle=False)
    DATA_hs.append(t+'_hs.npy')

In [234]:
def VStacker(files_,t):
    
    #Lê dados do arquivo e remove janela
    Xc = np.load('DB_REPO/stacked/'+files_[0])
    
    #Aplica Hstack nos dados
    for j in range(0,len(files_)-1):
        Xci = np.load('DB_REPO/stacked/'+files_[j])
        Xc = np.vstack((Xc,Xci)) 
          
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    np.save('./DB_REPO/stacked/'+t+'_vs', Xc, allow_pickle=False)
    DATA_vs.append(t+'_vs.npy')

In [235]:
def HStackerF(files_,t):
    
    #Lê dados do arquivo e remove janela
    Xc = np.load('DB_REPO/stacked/'+files_[5])
    Xc = Xc[:35853] #Exclui amostras para igualar tamanho e possibilitar o vstack
    
    #Aplica Hstack nos dados
    for j in range(0,len(files_)-1):
        Xci = np.load('DB_REPO/stacked/'+files_[j])
        Xci = Xci[:35853] #Exclui amostras para igualar tamanho e possibilitar o vstack
        Xc = np.hstack((Xc.reshape(Xc.shape),Xci.reshape(Xci.shape)))
          
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    np.save('./DB_REPO/stacked/'+t+'_fs', Xc, allow_pickle=False)
    DATA_fs.append(t+'_fs.npy')

In [236]:
#Carregando arquivos
import commands

n_features = 4
A = commands.getoutput('ls DB_REPO/preproc')
files_ = A.split('\n')
DATA_hs = []

for i in range(0,len(files_),n_features):
    HStacker(files_[i:i+n_features],files_[i:i+n_features][0].split('-')[0]+str(i/n_features))
    #print (files_[i:i+3],files_[0:3][0].split('-')[0]+str(i/3))

In [237]:
(files_[i:i+n_features],files_[i:i+n_features][0].split('-')[0]+str(i/n_features))

(['PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_fft.npy',
  'PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_filt.npy',
  'PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_rms.npy',
  'PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_std.npy'],
 'PK_Sergio Lamenza 83')

In [238]:
files_[i:i+n_features][0].split('-')[0]+str(i/n_features)

'PK_Sergio Lamenza 83'

In [239]:
files_[0:][0].split('-')[0]

'CT_Antonia Perpetuo Borges '

In [240]:
files_[0:4]

['CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_fft.npy',
 'CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_filt.npy',
 'CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_rms.npy',
 'CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_std.npy']

In [242]:
for i in range(0,len(files_),4):
    r = files_[i:i+4]
    
    print (((np.load('DB_REPO/preproc/'+r[0])).shape[0],(np.load('DB_REPO/preproc/'+r[1])).shape[0],
            (np.load('DB_REPO/preproc/'+r[2])).shape[0],(np.load('DB_REPO/preproc/'+r[3])).shape[0]))

(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(14000, 13951, 13951, 14000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(22000, 21951, 21951, 22000)
(14000, 13951, 13951, 14000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(22000, 21951,

In [243]:
import re 

print  (re.findall('\d*\D+',(DATA_hs[0:5][0]))[0])



CT_Antonia Perpetuo Borges 


In [244]:
DATA_hs

['CT_Antonia Perpetuo Borges 0_hs.npy',
 'CT_Antonia Perpetuo Borges 1_hs.npy',
 'CT_Antonia Perpetuo Borges 2_hs.npy',
 'CT_Francisco das Chagas dos Santos 3_hs.npy',
 'CT_Francisco das Chagas dos Santos 4_hs.npy',
 'CT_Francisco das Chagas dos Santos 5_hs.npy',
 'CT_Gelson Amantos de Souza  6_hs.npy',
 'CT_Gelson Amantos de Souza  7_hs.npy',
 'CT_Gelson Amantos de Souza  8_hs.npy',
 'CT_Geraldo Dias de Oliveira 9_hs.npy',
 'CT_Geraldo Dias de Oliveira 10_hs.npy',
 'CT_Geraldo Dias de Oliveira 11_hs.npy',
 'CT_Iraci Rodrigues Soares 12_hs.npy',
 'CT_Iraci Rodrigues Soares 13_hs.npy',
 'CT_Iraci Rodrigues Soares 14_hs.npy',
 'CT_Joaquina de Sousa Freitas  15_hs.npy',
 'CT_Joaquina de Sousa Freitas  16_hs.npy',
 'CT_Joaquina de Sousa Freitas  17_hs.npy',
 'CT_Jose Martins de Sousa  18_hs.npy',
 'CT_Jose Martins de Sousa  19_hs.npy',
 'CT_Jose Martins de Sousa  20_hs.npy',
 'CT_Larcir Iria da Conceicao Santos 21_hs.npy',
 'CT_Larcir Iria da Conceicao Santos 22_hs.npy',
 'CT_Larcir Iria d

In [245]:
Xc = np.load('DB_REPO/stacked/PK_Arthur Araujo Filho 42_hs.npy')

In [246]:
dfn = pd.DataFrame(data=Xc)
dfn.head()

0         1         2         3           4          5           6   \
0  0.022327 -0.001221  0.186603  0.017612  175.605680  10.642500  591.227547   
1  0.019318 -0.001221  0.186603  0.015084   75.150074   4.178106  409.107829   
2  0.025335  0.000705  0.173561  0.024354  120.955158   1.798792  822.637772   
3  0.016310 -0.000258  0.173561  0.013398   76.398090   2.421141  242.650976   
4  0.022327  0.000224  0.199644  0.015926   46.518462   1.619408  259.992532   

          7         8         9         10        11        12        13  \
0  50.701473  0.000375  0.000054 -0.002637  0.000031  0.024166  0.004446   
1  23.394437  0.000290  0.000073 -0.001959 -0.000076  0.024076  0.004444   
2  21.162750  0.000312  0.000108 -0.001131 -0.000242  0.023946  0.004452   
3  33.787073  0.000597  0.000165 -0.000066 -0.000442  0.023946  0.004480   
4  20.320969  0.001167  0.000239  0.001152 -0.000584  0.024612  0.004537   

         14        15  
0  0.228747  0.011801  
1  0.229715  0.011622  
2  0.230168  0.011512  
3  0.231609  0.011092  
4  0.232767  0.011054

In [247]:
Xc.shape

(11951, 16)

In [248]:
len(DATA_hs)

84

In [249]:
#Carregando arquivos
DATA_vs = []

for i in range(0,len(DATA_hs),3):
    VStacker(DATA_hs[i:i+3],(re.findall('\d*\D+',DATA_hs[i:i+3][0])[0]+str(i/3)))

In [250]:
len(DATA_vs)

28

In [251]:
Xc = np.load('DB_REPO/stacked/PK_Arthur Araujo Filho 42_hs.npy')
dfn = pd.DataFrame(data=Xc)
#dfn.head()

Xc[:35853].shape

(11951, 16)

for i in range(0,len(DATA_vs),1):
    print min(np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][0]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][1]).shape[0],
              np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][2]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][3]).shape[0],
              np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][4]).shape[0], np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][5]).shape[0])

i= 6
print min(np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][0]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][1]).shape[0],
           np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][2]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][3]).shape[0],
           np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][4]).shape[0], np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][5]).shape[0])

In [252]:
DATA_vs

['CT_Antonia Perpetuo Borges 0_vs.npy',
 'CT_Francisco das Chagas dos Santos 1_vs.npy',
 'CT_Gelson Amantos de Souza  2_vs.npy',
 'CT_Geraldo Dias de Oliveira 3_vs.npy',
 'CT_Iraci Rodrigues Soares 4_vs.npy',
 'CT_Joaquina de Sousa Freitas  5_vs.npy',
 'CT_Jose Martins de Sousa  6_vs.npy',
 'CT_Larcir Iria da Conceicao Santos 7_vs.npy',
 'CT_Magda Aparecida Ferreira 8_vs.npy',
 'CT_Manuel Saraiva da Rocha 9_vs.npy',
 'CT_Maria das Gracas Vieira 10_vs.npy',
 'CT_Maria Santana Estevam Moreira 11_vs.npy',
 'CT_Maria Veira Borges 12_vs.npy',
 'CT_Zenobia Barroso Feitosa 13_vs.npy',
 'PK_Arthur Araujo Filho 14_vs.npy',
 'PK_Boanerges Ramos Cunha 15_vs.npy',
 'PK_Carmino Pereira da Silva 16_vs.npy',
 'PK_Diniz Esmeraldo 17_vs.npy',
 'PK_Donia Aires 18_vs.npy',
 'PK_Fernando Francisco da Silva 19_vs.npy',
 'PK_Gastao Goncalves da Silva 20_vs.npy',
 'PK_Ismael Rodrigues do Carmo 21_vs.npy',
 'PK_Ivanildes Barbosa 22_vs.npy',
 'PK_Jose Fernandes Araujo 23_vs.npy',
 'PK_Jose Francisto Fontineli 

DATA_vs[78:78+6]

DATA_fs = []

for i in range(0,len(DATA_vs),6):
    HStackerF(DATA_vs[i:i+6],(re.findall('\d*\D+',DATA_vs[i:i+6][0])[0]+str(i/6)))

DATA_fs

In [272]:
def proccess_data(x,DATA,LABEL,NUM_SAMPLES = 50):
    
    Xc = np.load("DB_REPO/stacked/"+x)
    print '\nXc initial shape ', Xc.shape
    
    #Principal component analysis
    #pca = PCA(n_components=3)
    #pca.fit(Xc)
    #Xc = pca.transform(Xc)
    #print pca.explained_variance_ratio_
    #print ('Entropia: %3f' % (sum(pca.explained_variance_ratio_)))
    #print '\nXc PCA shape ', Xc.shape
    
    #Create temporal serie
    Xc = generate_envelope(Xc, NUM_SAMPLES)
    print 'Xc temporal-serie shape ', Xc.shape
    
    #Labeling the PKS level    
    C = (np.ones(len(Xc))*LABEL).reshape((len(Xc),1))
    Xc = np.hstack((Xc.reshape(Xc.shape),C.reshape((len(Xc),1)))) 
    print 'Xc labeled shape ', Xc.shape
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    #np.save('./DB_REPO/labeled/'+x[:2]+'_tsl', Xc, allow_pickle=False)
    np.save('./DB_REPO/labeled/'+x[:-4]+'_'+str(y), Xc, allow_pickle=False)
    #print ''+x[:2]+'_tsl'
    print x[:-4]+'_'+str(y)
    #DATA.append(x[:2]+'_tsl.npy')
    DATA.append(x[:-4]+'_'+str(y))

---

#Validação
start = timeit.default_timer()

NUM_SAMPLES = 500
l = [1]
DATA = []

for x, y in zip(DATA_hs, l):
    proccess_data(x,DATA,NUM_SAMPLES,y)
    
stop = timeit.default_timer()
ti = stop - start 
print ('Tempo total para o envelopamento: %.2f min' % (ti/60))

DATA

Xc = np.load('DB_INTENC/preproc/labeled/p4_hs.npy_tsl.npy')
dfn = pd.DataFrame(data=Xc)
dfn.head()

---

In [273]:
#X = 'F'
#Y = 'F'
#Z = 'F'
#l = [1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1,
# X, X, X, X, X, X, 15, 15, 15, 15, 15, 15, Y, Y, Y, Y, Y, Y, 15, 15, 15, 15, 15, 15, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
# 3, 3, 3, 3, 3, 3, Z, Z, Z, Z, Z, Z, 1, 1, 1, 1, 1, 1, 4, 4, 4, 4, 4, 4, 1, 1, 1, 1, 1, 1]

#l = [1, 2, 15, 15, 2, 1, 15, 15, 1, 1, 3, 1, 4, 1]
l = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
#l = [2, 15, 4, 1, 0, 0, 15]
for x, y in zip(DATA_vs, l):
    print (x[:-4]+'_'+str(y))

CT_Antonia Perpetuo Borges 0_vs_0
CT_Francisco das Chagas dos Santos 1_vs_0
CT_Gelson Amantos de Souza  2_vs_0
CT_Geraldo Dias de Oliveira 3_vs_0
CT_Iraci Rodrigues Soares 4_vs_0
CT_Joaquina de Sousa Freitas  5_vs_0
CT_Jose Martins de Sousa  6_vs_0
CT_Larcir Iria da Conceicao Santos 7_vs_0
CT_Magda Aparecida Ferreira 8_vs_0
CT_Manuel Saraiva da Rocha 9_vs_0
CT_Maria das Gracas Vieira 10_vs_0
CT_Maria Santana Estevam Moreira 11_vs_0
CT_Maria Veira Borges 12_vs_0
CT_Zenobia Barroso Feitosa 13_vs_0
PK_Arthur Araujo Filho 14_vs_1
PK_Boanerges Ramos Cunha 15_vs_1
PK_Carmino Pereira da Silva 16_vs_1
PK_Diniz Esmeraldo 17_vs_1
PK_Donia Aires 18_vs_1
PK_Fernando Francisco da Silva 19_vs_1
PK_Gastao Goncalves da Silva 20_vs_1
PK_Ismael Rodrigues do Carmo 21_vs_1
PK_Ivanildes Barbosa 22_vs_1
PK_Jose Fernandes Araujo 23_vs_1
PK_Jose Francisto Fontineli  24_vs_1
PK_Marcilio Pontinho de Rezende 25_vs_1
PK_Osvaldir Borborema 26_vs_1
PK_Sergio Lamenza 27_vs_1


In [274]:
for x, y in zip(DATA_vs[0:1], l[0:1]):
    print (x[:-4]+'_'+str(y))

CT_Antonia Perpetuo Borges 0_vs_0


In [275]:
start = timeit.default_timer()

DATA = []

for x, y in zip(DATA_vs[0:1], l[0:1]):
    proccess_data(x,DATA,y)
    
stop = timeit.default_timer()
ti = stop - start 
print ('Tempo total para o envelopamento: %.2f min' % (ti/60))


Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Antonia Perpetuo Borges 0_vs_0
Tempo total para o envelopamento: 25.67 min


In [265]:
#Envelope size.

#Grupo total
start = timeit.default_timer()

DATA = []

for x, y in zip(DATA_vs, l):
    proccess_data(x,DATA,y)
    
stop = timeit.default_timer()
ti = stop - start 
print ('Tempo total para o envelopamento: %.2f min' % (ti/60))


Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35354, 8000)
Xc labeled shape  (35354, 8001)
CT_Antonia Perpetuo Borges 0_vs_0

Xc initial shape  (35853, 16)


KeyboardInterrupt: 

#Não foram incluidas as amostras p4 e c3, serão utilizadas para validação!

In [196]:
DATA[2:-2]

['CT_Gelson Amantos de Souza  2_vs_0',
 'CT_Geraldo Dias de Oliveira 3_vs_0',
 'CT_Iraci Rodrigues Soares 4_vs_0',
 'CT_Joaquina de Sousa Freitas  5_vs_0',
 'CT_Jose Martins de Sousa  6_vs_0',
 'CT_Larcir Iria da Conceicao Santos 7_vs_0',
 'CT_Magda Aparecida Ferreira 8_vs_0',
 'CT_Manuel Saraiva da Rocha 9_vs_0',
 'CT_Maria das Gracas Vieira 10_vs_0',
 'CT_Maria Santana Estevam Moreira 11_vs_0',
 'CT_Maria Veira Borges 12_vs_0',
 'CT_Zenobia Barroso Feitosa 13_vs_0',
 'PK_Arthur Araujo Filho 14_vs_1',
 'PK_Boanerges Ramos Cunha 15_vs_1',
 'PK_Carmino Pereira da Silva 16_vs_1',
 'PK_Diniz Esmeraldo 17_vs_1',
 'PK_Donia Aires 18_vs_1',
 'PK_Fernando Francisco da Silva 19_vs_1',
 'PK_Gastao Goncalves da Silva 20_vs_1',
 'PK_Ismael Rodrigues do Carmo 21_vs_1',
 'PK_Ivanildes Barbosa 22_vs_1',
 'PK_Jose Fernandes Araujo 23_vs_1',
 'PK_Jose Francisto Fontineli  24_vs_1',
 'PK_Marcilio Pontinho de Rezende 25_vs_1']

In [197]:
len(DATA[2:-2])

24

In [198]:
DATA[-1]

'PK_Sergio Lamenza 27_vs_1'

---

## Shape data

In [199]:
Xc = np.load("DB_REPO/labeled/"+DATA[0]+'.npy')
for i in DATA[2:-2]:
    Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+i+'.npy'))) 

### Split tipo 1

#Dividindo os dados em sets de treino, test e validação
dfn = pd.DataFrame(data=Xc)
train, validate, test = np.split(dfn.sample(frac=1), [int(.6*len(dfn)), int(.8*len(dfn))])

np.save('./DB_INTENC/preproc/labeled/train', train, allow_pickle=False)
np.save('./DB_INTENC/preproc/labeled/validate', validate, allow_pickle=False)
np.save('./DB_INTENC/preproc/labeled/test', test, allow_pickle=False)

Xc_train = np.load("DB_INTENC/preproc/labeled/train.npy")

X_train = Xc_train[:,0:(Xc_train.shape[1]-1)]

yz = Xc_train[:,[(Xc_train.shape[1]-1)]]
y_train = np.array([])
for i in range(len(yz)):
    y_train = np.hstack((y_train,yz[i]))
    
lab_enc = preprocessing.LabelEncoder()
y_train = lab_enc.fit_transform(y_train)

X_train.shape, y_train.shape

#Xc_validate = np.load("DB_INTENC/preproc/labeled/validate.npy")
Xc_validate = Xc

X_validate = Xc_validate[:,0:(Xc_validate.shape[1]-1)]

yz = Xc_validate[:,[(Xc_validate.shape[1]-1)]]
y_validate = np.array([])
for i in range(len(yz)):
    y_validate = np.hstack((y_validate,yz[i]))
    
#lab_enc = preprocessing.LabelEncoder()
#y_validate = lab_enc.fit_transform(y_validate)

X_validate.shape, y_validate.shape

Xc_test= np.load("DB_INTENC/preproc/labeled/validate.npy")

X_test = Xc_test[:,0:(Xc_test.shape[1]-1)]

yz = Xc_test[:,[(Xc_test.shape[1]-1)]]
y_test = np.array([])
for i in range(len(yz)):
    y_test = np.hstack((y_test,yz[i]))
    
lab_enc = preprocessing.LabelEncoder()
y_test = lab_enc.fit_transform(y_test)

X_test.shape, y_test.shape

### Split tipo 2

In [200]:
print DATA[0]
print DATA[1]
print DATA[-1]
print DATA[-2]

CT_Antonia Perpetuo Borges 0_vs_0
CT_Francisco das Chagas dos Santos 1_vs_0
PK_Sergio Lamenza 27_vs_1
PK_Osvaldir Borborema 26_vs_1


In [201]:
for i in DATA[1:3]:
    print i

CT_Francisco das Chagas dos Santos 1_vs_0
CT_Gelson Amantos de Souza  2_vs_0


In [202]:
#Separando set de dados X, e set de labels y - Validação DATA[5] e DATA[6]
Xc = np.load("DB_REPO/labeled/"+DATA[0]+'.npy')

#for i in DATA[1:3]:
#    Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+i))) 

Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+DATA[1]+'.npy')))
Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+DATA[-1]+'.npy')))
Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+DATA[-2]+'.npy')))
    
X_validate = Xc[:,0:(Xc.shape[1]-1)]

yz = Xc[:,[(Xc.shape[1]-1)]]
y_validate = np.array([])
for i in range(len(yz)):
    y_validate = np.hstack((y_validate,yz[i]))
    
X_validate.shape, y_validate.shape

((143412, 3), (143412,))

In [208]:
X_validate_std = sc.transform(X_validate)

X_validate_std.shape, y_validate.shape

((143412, 3), (143412,))

In [204]:
#Separando set de dados X, e set de labels y - Treino e Teste
Xc = np.load("DB_REPO/labeled/"+DATA[3]+'.npy')
#for i in DATA[1:-2]:
for i in DATA[4:-2]:
    Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+i+'.npy'))) 
    
X = Xc[:,0:(Xc.shape[1]-1)]

yz = Xc[:,[(Xc.shape[1]-1)]]
y = np.array([])
for i in range(len(yz)):
    y = np.hstack((y,yz[i]))
    
X.shape, y.shape

((872619, 3), (872619,))

In [205]:
np.unique(y)

array([ 0.,  1.])

In [206]:
#Dividindo os dados em sets de treino e test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

#X_test, X_validate, y_test, y_validate = train_test_split(X, y, test_size=0.5, random_state=1)

In [207]:
#Aplica normalizador P.S.: Verificar necessidade pois os dados já foram normalizados anteriormente
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)
X_validate_std = sc.transform(X_validate)

print X_train_std.shape, X_test_std.shape, X_validate_std.shape, y_train.shape, y_test.shape, y_validate.shape

(610833, 3) (261786, 3) (143412, 3) (610833,) (261786,) (143412,)


---

## Classifiers

### Random Forest

In [209]:
#Training and test
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=100, n_jobs = 4)

print 'Training time'
%time rfc = rfc.fit(X_train_std, y_train)

print''

print 'Predict time'
%time y_pred = rfc.predict(X_test_std)

print''

from sklearn.metrics import accuracy_score
print ('ClassifyRF accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Training time
CPU times: user 5min 10s, sys: 4.18 s, total: 5min 14s
Wall time: 1min 26s

Predict time
CPU times: user 21.8 s, sys: 16 ms, total: 21.8 s
Wall time: 5.63 s

ClassifyRF accuracy:---------->95.60 %


### Memória anterior
Training time
CPU times: user 4min 12s, sys: 1.13 s, total: 4min 13s
Wall time: 1min 20s

Predict time
CPU times: user 3.82 s, sys: 188 ms, total: 4 s
Wall time: 1.45 s

ClassifyRF accuracy:---------->98.35 %


#Salva a rede treinada para um arquivo
joblib.dump(rfc, 'DB_INTENC/preproc/labeled/rfc_split2.pkl') 

rfc = joblib.load('DB_INTENC/preproc/labeled/rfc.pkl')

In [210]:
#Validation
print 'Predict time'
%time y_pred = rfc.predict(X_validate)

print''

from sklearn.metrics import accuracy_score

result = ('ClassifyRF accuracy:---------->%.2f %%' % (accuracy_score(y_pred, y_validate)*100))

print result

Predict time
CPU times: user 2.61 s, sys: 4 ms, total: 2.61 s
Wall time: 859 ms

ClassifyRF accuracy:---------->52.72 %


### Memória anterior
Predict time
CPU times: user 2.62 s, sys: 4 ms, total: 2.63 s
Wall time: 849 ms

ClassifyRF accuracy:---------->52.39 %


Predict time
CPU times: user 1.37 s, sys: 216 ms, total: 1.59 s
Wall time: 591 ms

ClassifyRF accuracy:---------->22.04 %


In [211]:
#Validation
print 'Predict time'
%time y_pred = rfc.predict(X_validate_std)

print''

from sklearn.metrics import accuracy_score

result = ('ClassifyRF accuracy:---------->%.2f %%' % (accuracy_score(y_pred, y_validate)*100))

print result

Predict time
CPU times: user 9.48 s, sys: 16 ms, total: 9.49 s
Wall time: 2.67 s

ClassifyRF accuracy:---------->61.55 %


### Memória anterior
Predict time
CPU times: user 9.48 s, sys: 12 ms, total: 9.49 s
Wall time: 2.63 s

ClassifyRF accuracy:---------->61.41 %

Predict time
CPU times: user 2.17 s, sys: 104 ms, total: 2.27 s
Wall time: 810 ms

ClassifyRF accuracy:---------->18.04 %

---

## Support Vector Machine

In [65]:
from sklearn import svm, datasets

print 'Processing time SVM'
%time svm = svm.SVC(kernel='rbf', gamma=0.2, C=6, decision_function_shape='ovr').fit(X_train_std, y_train)

print ''

from sklearn.metrics import accuracy_score

print 'Predict time'  
%time y_pred = svm.predict(X_test_std)
from sklearn.metrics import accuracy_score
print ('SVM-RBF accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Processing time SVM
CPU times: user 1h 45min 48s, sys: 2.05 s, total: 1h 45min 50s
Wall time: 1h 45min 54s

Predict time
CPU times: user 11min 41s, sys: 48 ms, total: 11min 41s
Wall time: 11min 42s
SVM-RBF accuracy:---------->81.72 %


In [66]:
print 'Predict time'  
%time y_pred = svm.predict(X_validate_std)
from sklearn.metrics import accuracy_score
print ('SVM-RBF accuracy:---------->%.2f %%' % (accuracy_score(y_validate, y_pred)*100))

Predict time
CPU times: user 10min 56s, sys: 24 ms, total: 10min 56s
Wall time: 10min 56s
SVM-RBF accuracy:---------->6.63 %


## TESTS

In [79]:
from sklearn.ensemble import AdaBoostClassifier
print 'Processing time ADA'
%time ada = AdaBoostClassifier(n_estimators=600,learning_rate=0.3).fit(X_train_std, y_train)

print ''

print 'Predict time'  
%time y_pred = ada.predict(X_test_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Processing time ADA
CPU times: user 6min 55s, sys: 52 ms, total: 6min 55s
Wall time: 6min 55s

Predict time
CPU times: user 15.4 s, sys: 12 ms, total: 15.5 s
Wall time: 15.5 s
ADA accuracy:---------->59.37 %


In [80]:
print 'Predict time'  
%time y_pred = ada.predict(X_validate_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_validate, y_pred)*100))

Predict time
CPU times: user 13.9 s, sys: 0 ns, total: 13.9 s
Wall time: 13.9 s
ADA accuracy:---------->6.72 %


## 2

In [73]:
from sklearn.tree import DecisionTreeClassifier
print 'Processing time ADA'
%time ada = AdaBoostClassifier(DecisionTreeClassifier(max_depth=2),n_estimators=600,learning_rate=1).fit(X_train_std, y_train)

print ''

print 'Predict time'  
%time y_pred = ada.predict(X_test_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Processing time ADA
CPU times: user 13min 4s, sys: 288 ms, total: 13min 4s
Wall time: 13min 5s

Predict time
CPU times: user 15 s, sys: 0 ns, total: 15 s
Wall time: 15 s
ADA accuracy:---------->39.56 %


In [74]:
print 'Predict time'  
%time y_pred = ada.predict(X_validate_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_validate, y_pred)*100))

Predict time
CPU times: user 14.3 s, sys: 0 ns, total: 14.3 s
Wall time: 14.3 s
ADA accuracy:---------->10.59 %
